

% RECOMMENDED %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\documentclass[graybox]{svmult}

% choose options for [] as required from the list
% in the Reference Guide

%\usepackage{type1cm}        % activate if the above 3 fonts are
%                            % not available on your system
%
%\usepackage{makeidx}         % allows index generation
\usepackage{graphicx}        % standard LaTeX graphics tool
                             % when including figure files
\usepackage{multicol}        % used for the two-column index
\usepackage[bottom]{footmisc}% places footnotes at page bottom
\usepackage{framed} % vince may 17


\usepackage{newtxtext}       % 
\usepackage[varvw]{newtxmath}       % selects Times Roman as basic font

% see the list of further useful packages
% in the Reference Guide

%\makeindex             % used for the subject index
                       % please use the style svind.ist with
                       % your makeindex program

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{document}

\title*{Bioconductor's Computational Ecosystem for Genomic Data Science in Cancer}
\titlerunning{Bioconductor for Cancer Data Science}
% Use \titlerunning{Short Title} for an abbreviated version of
% your contribution title if the original one is too long
\author{Name of First Author\orcidID{0000-1111-2222-3333} and\\ Name of Second Author\orcidID{1111-2222-3333-4444}}
% Use \authorrunning{Short Title} for an abbreviated version of
% your contribution title if the original one is too long
\institute{Name of First Author \at Name, Address of Institute, \email{name@email.address}
\and Name of Second Author \at Name, Address of Institute \email{name@email.address}}
%
% Use the package "url.sty" to avoid
% problems with special characters
% used in your e-mail or web address
%
\maketitle

\abstract{
The Bioconductor project enters its third decade with over two 
thousand packages for genomic data science, over 100,000 annotation and 
experiment resources, and a global system for convenient distribution to 
researchers. Over 60,000 PubMed Central citations and terabytes of content 
shipped per month attest to the impact of the project 
on cancer genomic data science. This report provides an overview 
of cancer genomics resources in Bioconductor. After an overview 
of Bioconductor project principles, we address exploration 
of institutionally curated cancer genomics data such as TCGA. 
We then review genomic annotation and ontology resources 
relevant to cancer and then briefly survey analytical 
workflows addressing specific topics in cancer genomics. 
Concluding sections cover how new software and data 
resources are brought into the ecosystem and how the 
project is tackling needs for training of the research 
workforce. Bioconductor's strategies for supporting 
methods developers and researchers in cancer genomics 
are evolving along with experimental and computational 
technologies. All the tools described in this report 
are backed by regularly maintained learning resources 
that can be used locally or in cloud computing environments.
}



\section{Introduction}
\label{sec:1}

Computation is a central component of cancer genomics
research. Tumor sequencing is the basis of computational
investigation of mutational, epigenetic and immunologic
processes associated with cancer initiation and progression.
Numerous computational workflows have been produced to
profile tumor cell transcriptomes and proteomes.
New technologies promise to unite sequence-based
characterizations with digital histopathology,
ultimately driving efforts in molecule design
and evaluation to produce patient-centered treatments.

Bioconductor is an open source software project with
a 20 year history of uniting biostatisticians, bioinformaticians,
and genome researchers in the creation of an ecosystem
of data, annotation, and analysis resources for research
in genome-scale biology. This paper will review current
approaches of the project to advancing cancer genomics.
After a brief discussion of basic principles of the Bioconductor
project, we will present a ``top down'' survey of resources
useful for cancer bioinformatics. Primary sections address

\begin{itemize}
\tightlist
\item
  how to explore institutionally curated cancer genomics data
\item
  genomic annotation resources relevant to cancer genomics
\item
  analytical workflows
\item
  components for introducing new data or analyses
\item
  pedagogics and workforce development.
\end{itemize}

Final sections provide enumerations of software and data packages
tagged by their contributors as specifically relevant to cancer.

\section{Bioconductor principles}


\subsection{R packages and vignettes}\label{r-packages-and-vignettes}}

Software tools and data resources in Bioconductor are organized
into ``R packages''. These are collections of folders with data,
code (principally R functions), and documentation
following a protocol specified in
the
Writing R Extensions manual \cite{WRE}.  R packages have a DESCRIPTION file with metadata about
package contents and provenance. Package structure can be
checked for validity using the \texttt{R CMD check} facility.
Documentation of code and data can be programmatically
checked for existence and validity. The DESCRIPTION file
for a package specifies its version and
also gives precise definition of how an R package may
depend upon versions of other packages.

At its inception,
Bioconductor introduced a new approach to holistic package
documentation called ``vignette''.
Vignettes provide narrative and explanation interleaved with
executable code describing package operations.
While R function manual pages describe
the operation of individual functions,
vignettes illustrate the interoperation
of package components and provide motivation
for both package design but also context
for its use.

\subsection{R package repositories; repository evolution}\label{r-package-repositories-repository-evolution}}

Bioconductor software forms a coherent ecosystem that
can be checked for consistency of versions of all
packages available in a given installation of R.
Bioconductor packages may specify dependency on
other Bioconductor packages, or packages that are
available in the CRAN repository. Bioconductor does
not include packages with dependencies on ``github-only''
packages. Later in this paper we will provide details
on package quality assurance that provide a rationale
for this restriction.

Major updates to the R language occur annually, and
updates are preceded by careful assessment of effects of
language change on Bioconductor package operations. These effects
can be identified through changes in the output of R CMD check.
The Bioconductor ecosystem is updated twice a year, once
to coincide with update to R, and once about six months
later. The semianual updates reflect the need to track
developments in the fast-moving field of genomic data science.

\subsection{Package quality assessment; installation consistency}\label{package-quality-assessment-installation-consistency}}

The BiocCheck function is used to provide more
stringent assessment of package compliance with basic
principles of the Bioconductor ecosystem.

The BiocManager package provides for installing and updating package
and has functionality for verifying the coherence and version status
of the currently installed package collection.
This is important
in the context of a language and package ecosystem
that changes every six months, while analyses may
take years to complete. Tools for recreating past
package collections are available to assist in
reproducing outputs of prior analyses.

\subsection{Unifying assay and sample data: SummarizedExperiment and MultiAssayExperiment}\label{unifying-assay-and-sample-data-summarizedexperiment-and-multiassayexperiment}}

Most of the data from genome-scale experiments to be discussed
in this chapter are organized in special data containers
rooted in the concepts of the SummarizedExperiment class.
Briefly, assay data are thought of as occupying a \(G \times N\)
array, and sample level data occupy an \(N \times K\) table. The array
and the table are linked together in the SummarizedExperiment; see Figure \ref{fig:sesc}.

\begin{figure}
\includegraphics[width=0.8\linewidth,]{SEschema} \caption{SummarizedExperiment schematic.}\label{fig:sesc}
\end{figure}

Multiple representations of assay results may be managed in this
structure, but all assay arrays must have dimensions \(G \times N\).

For experiment collections in which the same samples are subjected
to multiple genome-scale assays, MultiAssayExperiment containers are used. See Figure \ref{fig:masc} for the layout.

\begin{figure}
\includegraphics[width=0.8\linewidth,]{MAEschema} \caption{MultiAssayExperiment schematic.}\label{fig:masc}
\end{figure}

Further details on these data structures will be provided in section \ref{class}.

\subsection{Downloading and caching cancer genomics data and annotations}\label{cache}}

Downloading and managing data from various online resources
can be excessively time consuming. Bioconductor encourages data caching for
increased efficiency and reproducibility. The caching data methods
employed in Bioconductor
allow analysis code to
concisely refer to data resources as needed, with minimal attention to how
data are stored, retrieved or transformed.
It allows for easy management and reuse of data that are on remote
servers or in cloud, storing source
location and providing information for data updates. The BiocFileCache
Bioconductor package handles data management from within R.

BiocFileCache is a general-use caching system but Bioconductor also provides
``Hubs'', AnnotationHub and ExperimentHub, to help distributed annotation or
experimental data hosted externally. Both AnnotationHub and ExperimentHub use
BiocFileCache to handle download and caching of data.

AnnotationHub provides a centralized repository of diverse genomic annotations,
facilitating easy access and integration into analyses. Researchers can
seamlessly retrieve information such as genomic features, functional
annotations, and variant data, streamlining the annotation process for their
analyses.

ExperimentHub extends this concept to experimental data. It serves as a
centralized hub for storing and sharing curated experiment-level datasets,
allowing researchers to access a wide range of experimental designs and
conditions. This cloud-based infrastructure enhances collaboration and promotes
the reproducibility of analyses across different laboratories.

The curatedTCGAData package provides some resources through
ExperimentHub, as do many other self-identified ``CancerData'' resources. Once the
ExperimentHub is loaded, it can be queried for terms of interest.

\begin{shaded}
\begin{verbatim}
library(ExperimentHub)
eh = ExperimentHub()
query(eh, "CancerData")
## ExperimentHub with 1742 records
## # snapshotDate(): 2024-04-29
## # $dataprovider: Eli and Edythe L. Broad Institute of Harvard and MIT, GEO, ...
## # $species: Homo sapiens, Mus musculus, NA
## # $rdataclass: SummarizedExperiment, RaggedExperiment, matrix, list, DFrame,...
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
## #   rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[["EH558"]]' 
## 
##            title                                
##   EH558  | ACC_CNASNP-20160128                  
##   EH559  | ACC_CNVSNP-20160128                  
##   EH560  | ACC_colData-20160128                 
##   EH561  | ACC_GISTIC_AllByGene-20160128        
##   EH562  | ACC_GISTIC_ThresholdedByGene-20160128
##   ...      ...                                  
##   EH8533 | tcga_transcript_counts               
##   EH8534 | target_rhabdoid_wgbs_hg19            
##   EH8567 | xenium_hs_breast_addon               
##   EH9482 | Capper_example_betas.rda             
##   EH9483 | GIMiCC_Library.rda    
\end{verbatim}
\end{shaded}

%\texttt{\{r useeh\} \textless{}!-\/- , fig.cap="ExperimentHub attachment, retrieval, query, and response when seeking cancer-related data.", message=FALSE\} -\/-\textgreater{} library(ExperimentHub) eh \textless{}- ExperimentHub() query(eh, "CancerData")}

Multiple terms can be used to narrow results before choosing a download.

\begin{shaded}
\begin{verbatim}
query(eh, c("CancerData", "esophageal"))
## ExperimentHub with 2 records}
## snapshotDate(): 2023-10-24}
## $dataprovider: University of California San Francisco}
## $species: Homo sapiens}
## $rdataclass: RangedSummarizedExperiment, data.frame}
## additional mcols(): taxonomyid, genome, description,}
##   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,}
##   rdatapath, sourceurl, sourcetype }
## retrieve records with, e.g., object[["EH8527"]]
##            title                           
##   EH8527 | cao_esophageal_wgbs_hg19        
##   EH8530 | cao_esophageal_transcript_counts
\end{verbatim}
\end{shaded}

Similarly AnnotationHub files can be downloaded for annotating data. For example,
the ensembl 110 release of gene and protein annotations are obtained with the
following:

\begin{shaded}
\begin{verbatim}
library(AnnotationHub)
ah = AnnotationHub()
query(ah, c("ensembl", "110", "Homo sapiens"))
#snapshotDate(): 2024-04-29
#AnnotationHub with 1 record
## snapshotDate(): 2024-04-29
## names(): AH113665
## $dataprovider: Ensembl
## $species: Homo sapiens
## $rdataclass: EnsDb
## $rdatadateadded: 2023-04-25
## $title: Ensembl 110 EnsDb for Homo sapiens
## $description: Gene and protein annotations for Homo sapiens based on Ensem...
## $taxonomyid: 9606
## $genome: GRCh38
## $sourcetype: ensembl
## $sourceurl: http://www.ensembl.org
## $sourcesize: NA
## $tags: c("110", "Annotation", "AnnotationHubSoftware", "Coverage",
##   "DataImport", "EnsDb", "Ensembl", "Gene", "Protein", "Sequencing",
##   "Transcript") 
## retrieve record with 'object[["AH113665"]]' 
\end{verbatim}
\end{shaded}

\input{inst}

\input{workf}

\input{spat}

\input{endit}

\pagebreak

\input{fig7}

\section{Acknowledgments}\label{acknowledgments}

This work was supported in part by NIH NCI 3U24CA180996-10S1, NHGRI 5U24HG004059-18, and NSF ACCESS allocation BIR190004.


%\input{newrefs}
\bibliographystyle{vjcunsrt}
\bibliography{biocmimb}

\end{document}
